### ISSUE TITLE
Health endpoint has CPU performance regression factor two in version v1.15.0

### ISSUE THREAD

**Issue by beyhan** [cloudfoundry/dns-release#52]
Steps to reproduce the issue on a machine with up to 4 CPUs and 6 GB RAM:

1. Deploy a bosh-lite with the steps from https://bosh.io/docs/bosh-lite/

2. Deploy zookeeper as described in https://bosh.io/docs/bosh-lite/#deploy with 10 instances instead of 5

3. Select a zookeeper instance and ssh into it.

4. Execute `top -p <bosh-dns-health-process-id>`

5. Execute dns queries from all other instances with `bosh is -d zookeeper | grep -v <instance-id> | grep -v smoke-tests | awk '{print $1}' | xargs -I {} -P 9 bosh -d zookeeper ssh {} -c "for i in {1..10}; do printf 'dig q-s3-i<instance-bosh-index>.zookeeper.*.*.bosh; echo %s;\\n' {0..20} | xargs -P 20 -I % sh -c '%'; done"`

Executing the described steps for the `bosh-dns` releases `v1.12.0` and `v1.15.0` I got following results for the CPU usage:

`bosh-dns` version `v1.12.0`
```
top - 11:59:15 up  1:36,  1 user,  load average: 2.07, 6.32, 7.56
Tasks:   1 total,   0 running,   1 sleeping,   0 stopped,   0 zombie
%Cpu(s): 50.8 us, 43.0 sy,  0.0 ni,  0.0 id,  0.0 wa,  0.0 hi,  6.2 si,  0.0 st
KiB Mem :  6103628 total,   352636 free,  3204944 used,  2546048 buff/cache
KiB Swap:  6103036 total,  6052348 free,    50688 used.  2671844 avail Mem 

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
   8252 vcap       5 -15  672876 158108   6744 S  17.7  2.6   0:47.53 bosh-dns-health

top - 11:59:18 up  1:36,  1 user,  load average: 14.48, 8.82, 8.36
Tasks:   1 total,   0 running,   1 sleeping,   0 stopped,   0 zombie
%Cpu(s): 46.8 us, 46.7 sy,  0.0 ni,  0.0 id,  0.0 wa,  0.0 hi,  6.5 si,  0.0 st
KiB Mem :  6103628 total,   362376 free,  3195292 used,  2545960 buff/cache
KiB Swap:  6103036 total,  6052348 free,    50688 used.  2681684 avail Mem 

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
   8252 vcap       5 -15  672876 158108   6744 S  22.0  2.6   0:48.19 bosh-dns-health

top - 11:59:21 up  1:36,  1 user,  load average: 14.48, 8.82, 8.36
Tasks:   1 total,   0 running,   1 sleeping,   0 stopped,   0 zombie
%Cpu(s): 49.0 us, 45.0 sy,  0.0 ni,  0.0 id,  0.0 wa,  0.0 hi,  6.0 si,  0.0 st
KiB Mem :  6103628 total,   325992 free,  3230824 used,  2546812 buff/cache
KiB Swap:  6103036 total,  6052348 free,    50688 used.  2645500 avail Mem 

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
   8252 vcap       5 -15  672876 158108   6744 S  21.3  2.6   0:48.83 bosh-dns-health

top - 11:59:24 up  1:36,  1 user,  load average: 23.73, 10.83, 9.01
Tasks:   1 total,   1 running,   0 sleeping,   0 stopped,   0 zombie
%Cpu(s): 51.5 us, 43.0 sy,  0.0 ni,  0.0 id,  0.0 wa,  0.0 hi,  5.4 si,  0.0 st
KiB Mem :  6103628 total,   321932 free,  3234444 used,  2547252 buff/cache
KiB Swap:  6103036 total,  6052348 free,    50688 used.  2641944 avail Mem 

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
   8252 vcap       5 -15  672876 158108   6744 R  18.6  2.6   0:49.39 bosh-dns-health

top - 11:59:27 up  1:36,  1 user,  load average: 24.87, 11.28, 9.17
Tasks:   1 total,   0 running,   1 sleeping,   0 stopped,   0 zombie
%Cpu(s): 59.2 us, 35.4 sy,  0.0 ni,  0.0 id,  0.0 wa,  0.0 hi,  5.3 si,  0.0 st
KiB Mem :  6103628 total,   427656 free,  3130796 used,  2545176 buff/cache
KiB Swap:  6103036 total,  6052348 free,    50688 used.  2748068 avail Mem 

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
   8252 vcap       5 -15  672876 158108   6744 S  17.6  2.6   0:49.92 bosh-dns-health

top - 11:59:30 up  1:36,  1 user,  load average: 24.87, 11.28, 9.17
Tasks:   1 total,   0 running,   1 sleeping,   0 stopped,   0 zombie
%Cpu(s): 33.7 us,  9.0 sy,  0.0 ni, 55.6 id,  0.2 wa,  0.0 hi,  1.6 si,  0.0 st
KiB Mem :  6103628 total,   463644 free,  3094524 used,  2545460 buff/cache
KiB Swap:  6103036 total,  6052348 free,    50688 used.  2784780 avail Mem 

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
   8252 vcap       5 -15  672876 158108   6744 S   0.7  2.6   0:49.94 bosh-dns-health

top - 11:59:33 up  1:36,  1 user,  load average: 23.04, 11.13, 9.13
Tasks:   1 total,   0 running,   1 sleeping,   0 stopped,   0 zombie
%Cpu(s):  0.4 us,  0.6 sy,  0.0 ni, 98.9 id,  0.0 wa,  0.0 hi,  0.1 si,  0.0 st
KiB Mem :  6103628 total,   463768 free,  3094384 used,  2545476 buff/cache
KiB Swap:  6103036 total,  6052348 free,    50688 used.  2784904 avail Mem 

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
   8252 vcap       5 -15  672876 158108   6744 S   0.0  2.6   0:49.94 bosh-dns-health

```

`bosh-dns` version `v1.15.0`

```
top - 19:00:55 up  8:37,  1 user,  load average: 12.01, 7.47, 4.99
Tasks:   1 total,   0 running,   1 sleeping,   0 stopped,   0 zombie
%Cpu(s):  5.0 us,  2.6 sy,  0.0 ni, 91.9 id,  0.1 wa,  0.0 hi,  0.4 si,  0.0 st
KiB Mem :  6103628 total,   222404 free,  3478396 used,  2402828 buff/cache
KiB Swap:  6103036 total,  6019836 free,    83200 used.  2828168 avail Mem 

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
  77074 vcap       5 -15  682896 167684   8184 S   0.0  2.7   3:29.36 bosh-dns-health

top - 19:00:58 up  8:37,  1 user,  load average: 11.05, 7.35, 4.97
Tasks:   1 total,   0 running,   1 sleeping,   0 stopped,   0 zombie
%Cpu(s):  4.3 us,  3.0 sy,  0.0 ni, 92.3 id,  0.1 wa,  0.0 hi,  0.3 si,  0.0 st
KiB Mem :  6103628 total,   208176 free,  3492616 used,  2402836 buff/cache
KiB Swap:  6103036 total,  6019836 free,    83200 used.  2814024 avail Mem 

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
  77074 vcap       5 -15  682896 167684   8184 S   2.0  2.7   3:29.42 bosh-dns-health

top - 19:01:02 up  8:37,  1 user,  load average: 14.01, 8.02, 5.20
Tasks:   1 total,   0 running,   1 sleeping,   0 stopped,   0 zombie
%Cpu(s): 52.5 us, 12.4 sy,  0.0 ni, 31.4 id,  0.1 wa,  0.0 hi,  3.6 si,  0.0 st
KiB Mem :  6103628 total,   117592 free,  3610028 used,  2376008 buff/cache
KiB Swap:  6103036 total,  6019580 free,    83456 used.  2696692 avail Mem 

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
  77074 vcap       5 -15  682896 167360   7860 S   8.0  2.7   3:29.68 bosh-dns-health

top - 19:01:05 up  8:38,  1 user,  load average: 14.01, 8.02, 5.20
Tasks:   1 total,   1 running,   0 sleeping,   0 stopped,   0 zombie
%Cpu(s): 65.4 us, 30.8 sy,  0.0 ni,  0.0 id,  0.0 wa,  0.0 hi,  3.9 si,  0.0 st
KiB Mem :  6103628 total,   198220 free,  3579368 used,  2326040 buff/cache
KiB Swap:  6103036 total,  6018812 free,    84224 used.  2727188 avail Mem 

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
  77074 vcap       5 -15  682896 166748   7268 R  33.2  2.7   3:30.71 bosh-dns-health

top - 19:01:08 up  8:38,  1 user,  load average: 22.34, 9.85, 5.80
Tasks:   1 total,   1 running,   0 sleeping,   0 stopped,   0 zombie
%Cpu(s): 62.6 us, 34.5 sy,  0.0 ni,  0.0 id,  0.0 wa,  0.0 hi,  2.8 si,  0.0 st
KiB Mem :  6103628 total,   177936 free,  3597428 used,  2328264 buff/cache
KiB Swap:  6103036 total,  6018812 free,    84224 used.  2706688 avail Mem 

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
  77074 vcap       5 -15  682896 166748   7268 R  34.2  2.7   3:31.74 bosh-dns-health

top - 19:01:11 up  8:38,  1 user,  load average: 22.34, 9.85, 5.80
Tasks:   1 total,   0 running,   1 sleeping,   0 stopped,   0 zombie
%Cpu(s): 48.0 us, 46.8 sy,  0.0 ni,  0.0 id,  0.0 wa,  0.0 hi,  5.2 si,  0.0 st
KiB Mem :  6103628 total,   198500 free,  3576056 used,  2329072 buff/cache
KiB Swap:  6103036 total,  6018812 free,    84224 used.  2727468 avail Mem 

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
  77074 vcap       5 -15  682896 166748   7268 S  31.2  2.7   3:32.68 bosh-dns-health

top - 19:01:14 up  8:38,  1 user,  load average: 33.60, 12.39, 6.65
Tasks:   1 total,   0 running,   1 sleeping,   0 stopped,   0 zombie
%Cpu(s): 52.5 us, 42.5 sy,  0.0 ni,  0.0 id,  0.0 wa,  0.0 hi,  5.1 si,  0.0 st
KiB Mem :  6103628 total,   183288 free,  3589576 used,  2330764 buff/cache
KiB Swap:  6103036 total,  6018812 free,    84224 used.  2712436 avail Mem 

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
  77074 vcap       5 -15  682896 166748   7268 S  32.2  2.7   3:33.65 bosh-dns-health

top - 19:01:17 up  8:38,  1 user,  load average: 41.40, 14.36, 7.32
Tasks:   1 total,   1 running,   0 sleeping,   0 stopped,   0 zombie
%Cpu(s): 52.2 us, 42.9 sy,  0.0 ni,  0.0 id,  0.0 wa,  0.0 hi,  4.9 si,  0.0 st
KiB Mem :  6103628 total,   179356 free,  3589836 used,  2334436 buff/cache
KiB Swap:  6103036 total,  6018812 free,    84224 used.  2708456 avail Mem 

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
  77074 vcap       5 -15  682896 166748   7268 R  31.9  2.7   3:34.62 bosh-dns-health

top - 19:01:20 up  8:38,  1 user,  load average: 41.40, 14.36, 7.32
Tasks:   1 total,   1 running,   0 sleeping,   0 stopped,   0 zombie
%Cpu(s): 81.3 us, 15.9 sy,  0.0 ni,  0.3 id,  0.0 wa,  0.0 hi,  2.5 si,  0.0 st
KiB Mem :  6103628 total,   196004 free,  3571032 used,  2336592 buff/cache
KiB Swap:  6103036 total,  6018812 free,    84224 used.  2727216 avail Mem 

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
  77074 vcap       5 -15  682896 166748   7268 R  44.9  2.7   3:35.97 bosh-dns-health

top - 19:01:23 up  8:38,  1 user,  load average: 43.45, 15.24, 7.64
Tasks:   1 total,   0 running,   1 sleeping,   0 stopped,   0 zombie
%Cpu(s): 81.7 us, 16.9 sy,  0.0 ni,  0.7 id,  0.0 wa,  0.0 hi,  0.7 si,  0.0 st
KiB Mem :  6103628 total,   277068 free,  3485664 used,  2340896 buff/cache
KiB Swap:  6103036 total,  6018812 free,    84224 used.  2809992 avail Mem 

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
  77074 vcap       5 -15  682896 166748   7268 S  42.3  2.7   3:37.24 bosh-dns-health

top - 19:01:26 up  8:38,  1 user,  load average: 43.45, 15.24, 7.64
Tasks:   1 total,   0 running,   1 sleeping,   0 stopped,   0 zombie
%Cpu(s): 63.3 us, 25.9 sy,  0.0 ni, 10.4 id,  0.0 wa,  0.0 hi,  0.4 si,  0.0 st
KiB Mem :  6103628 total,   275312 free,  3485632 used,  2342684 buff/cache
KiB Swap:  6103036 total,  6018812 free,    84224 used.  2808436 avail Mem 

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
  77074 vcap       5 -15  682896 166748   7268 S 154.5  2.7   3:41.89 bosh-dns-health
```

**Comment by cjnosal** [comment]
I'm having trouble reading the top output - it looks like the 1.15 section is prepended with the 1.12 outputs, but the remainder of the 1.15 section is missing the bosh-dns-health pid?

Was this tested against the 1.15 tag, or head of master?

**Comment by cjnosal** [comment]
I see peak user CPU at ~78 for 1.15 vs ~59 for 1.12, and peak system CPU ~46 for both in your output?

**Comment by beyhan** [comment]
@xtreme-conor-nosal I executed the test for `v1.15.0` again and updated the section. The tests for `v1.15.0` were done with the tag. I didn't try master. I tried `v1.15.0` with [this fix](https://github.com/cloudfoundry/bosh-dns-release/commit/0234da7db899f5e86004518546043a44c2e4e68a), which performs much better compared to `v1.15.0`. 

Performance tests are really hard, but what we observe in our dashboards is that the average CPU load has doubled. If you look into the average output of top, there you can also see that it is doubled. I'm doing CPU profiling for the health server.  Let's see whether I can identify a root cause. Do you have any idea, what could be the root cause?

**Comment by beyhan** [comment]
I'm uploading the CPU profiles for `v1.15.0` and `v1.12.0`. It looks like the version `v1.15.0` uses `TLS 1.3`, which seems to use more CPU. Is my understanding correct? 
 
[profile-bosh-health-v1.12.0.pdf](https://github.com/cloudfoundry/bosh-dns-release/files/3873562/profile-bosh-health-v1.12.0.pdf)
[profile-bosh-health-v1.15.0.pdf](https://github.com/cloudfoundry/bosh-dns-release/files/3873563/profile-bosh-health-v1.15.0.pdf)

**Comment by cjnosal** [comment]
We set a minimum TLS of 1.2 via https://github.com/pivotal-cf/paraphernalia/blob/master/secure/tlsconfig/config.go

Our golang dependency was updated to 1.13, which includes support for TLS 1.3, so I would expect it to use the stronger protocol. 

A quick googling suggests TLS 1.3 uses less CPU time and fewer round-trips to handshake. It's possible that the higher peak CPU is because more sessions can be accepted in a smaller time frame, with threads spending less time blocked on I/O. In other words, more resources are used because efficiency has increased.

The provided test script tries to churn through all the parallel requests as quickly as possible, so if there was no I/O, a spike to 100% usage would be expected. Maybe a valuable comparison scenario is to test is the average load when 1.12 and 1.15 receive requests at a constant rate?

**Comment by beyhan** [comment]
I agree that the provided test is not optimal, but it shows the symptoms of high CPU usage in `v1.15.0`. The `bosh-health` server in version `v1.15.0` can't even handle the load generated by this test and at some point `TLS handshake error from 10.0.65.98:53618: EOF` occurs in the `health` server logs, which doesn't happen with version `v1.120`.

The dashboard billow shows the CPU usage of the CF API nodes on one of our CF landscapes. You can see the spike of the CPU usage, when we introduced the `bosh-dns` in version `v1.15.0`.  At the end we had to downgrade the `bosh-dns` version to `v1.12.0`, because we experienced many TLS handshake errors with `connection reset by peer`. After downgrading to `v1.12.0`, the CPU usage went down and we don't experience TLS errors any more. The load on this CF landscape didn't change during that time.
<img width="2532" alt="overall-cpu-usage-cf-api" src="https://user-images.githubusercontent.com/517762/69427141-301c3380-0d2f-11ea-9a73-f4c0020a9391.png">

I'm not sure whether we can blame TLS 1.3 for this, but the CPU profiling shows that TLS handshake consumes ~80% of the CPU time in the `bosh-health` server.

**Comment by beyhan** [comment]
I modified `v1.15.0` to use the golang package used by `v1.12.0` and executed the reproduction test described in my first comment. I can confirm that with this change the CPU usage of version `v1.15.0` was the same as version `v1.12.0`.

**Comment by aashah** [comment]
We recently heard other feedback related to bosh-dns 1.15, CPU usage, and potentially the root cause being the golang bump.

We submitted a [commit](https://github.com/cloudfoundry/bosh-dns-release/commit/192bcf5ba1b96e63ef05e05e57715579d41013cf) that hopefully addresses the issue by pinning the TLS version to 1.2. When I get a chance, I'll try to run through the reproduction steps you gave with an updated build. I'll share the candidate build with you, and maybe you can help check as well.

**Comment by cjnosal** [comment]
@beyhan can you test this candidate in your environment? 
https://s3.amazonaws.com/bosh-dns-release-candidate-tarballs/bosh-dns-dev-release.tgz?versionId=ZbYBT5LysldT1zzDZYpZf80Y2BAT4msy

**Comment by beyhan** [comment]
@xtreme-conor-nosal is the release candidate based on current master branch state? We will test and give feedback.

**Comment by cjnosal** [comment]
Yes, built from 7bc1fed014151396a239c97b251e10268b5bec55

**Comment by openstack-cpi** [comment]
Thanks! We'll test it and expect feedback next week.

**Comment by friegger** [comment]
Thanks! We'll test it and expect feedback in the coming week.

**Comment by cjnosal** [comment]
We're aiming to cut a new release today with these changes. 

I've also determined that TLS 1.3 on its own doesn't account for the load. Testing against a dummy mTLS server I couldn't reproduce the higher CPU usage with TLS 1.3 vs 1.2. However, after tidying up the TLS client and server configurations used by bosh-dns (e.g. enabling keepalives, setting read/write timeouts, etc) the cpu overhead of 1.3 disappeared.

**Comment by beyhan** [comment]
@xtreme-conor-nosal Thanks for investigating this further and tiding the client and server configuration in https://github.com/cloudfoundry/bosh-dns-release/tree/tls-refactor

**Comment by beyhan** [comment]
We can conform that with version `v1.16.0` we don't experience CPU issues as with version `v1.15.0`.

**Comment by beyhan** [comment]
Do you want to create a new story for the `TLS1.3` refactoring or track it with this issue?

**Comment by aashah** [comment]
> We can conform that with version v1.16.0 we don't experience CPU issues as with version v1.15.0.

That's good to hear!

> Do you want to create a new story for the TLS1.3 refactoring or track it with this issue?

There is a chore for that here - https://www.pivotaltracker.com/story/show/170158998

Closing this issue as the immediate regression has been resolved.
